prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -1781,48 +1776,56 @@ layoutWadlerLeijen
         Nest j x        -> let !ij = i+j in best nl cc (Cons ij x ds)
         Union x y       -> let x' = best nl cc (Cons i x ds)
                                y' = best nl cc (Cons i y ds)
-                           in selectNicer fittingPredicate nl cc x' y'
+                           in selectNicer nl cc x' y'
         Column f        -> best nl cc (Cons i (f cc) ds)
         WithPageWidth f -> best nl cc (Cons i (f pWidth) ds)
         Nesting f       -> best nl cc (Cons i (f i) ds)
         Annotated ann x -> SAnnPush ann (best nl cc (Cons i x (UndoAnn ds)))
 
     selectNicer
-        :: FittingPredicate ann
-        -> Int           -- ^ Current nesting level
+        :: Int           -- ^ Current nesting level
         -> Int           -- ^ Current column
         -> SimpleDocStream ann -- ^ Choice A. Invariant: first lines should not be longer than B's.
         -> SimpleDocStream ann -- ^ Choice B.
         -> SimpleDocStream ann -- ^ Choice A if it fits, otherwise B.
-    selectNicer (FittingPredicate fits) lineIndent currentColumn x y
-      | fits pWidth minNestingLevel availableWidth x = x
-      | otherwise = y
+    selectNicer lineIndent currentColumn x y = case pWidth of
+        AvailablePerLine lineLength ribbonFraction
+          | fits pWidth minNestingLevel availableWidth x -> x
+          where
+            minNestingLevel =
+                -- See https://github.com/quchen/prettyprinter/issues/83.
+                if startsWithLine y
+                    -- y might be a (more compact) hanging layout. Let's check x
+                    -- thoroughly with the smaller lineIndent.
+                    then lineIndent
+                    -- y definitely isn't a hanging layout. Let's allow the first
+                    -- line of x to be checked on its own and format it consistently
+                    -- with subsequent lines with the same indentation.
+                    else currentColumn
+            availableWidth = min columnsLeftInLine columnsLeftInRibbon
+              where
+                columnsLeftInLine = lineLength - currentColumn
+                columnsLeftInRibbon = lineIndent + ribbonWidth - currentColumn
+                ribbonWidth =
+                    (max 0 . min lineLength . round)
+                        (fromIntegral lineLength * ribbonFraction)
+        Unbounded
+          -- TODO: Figure out why it's sufficient to check the first line.
```

Comment:
Thats a good observation!

Because of the algorithms recursive nature on `Union` all unions below this point have been evaluated to simpledoc and found to be non-fail already and. The only way fail is introduced is through `group` and thus through `Union` on the left side. And you cannot have a flattened document where it first has a `Line` and then a `Fail` without at least one `Union` above that `Fail`!

Reply:
Many thanks for the explanation! :)

Could you check the `Note [Detecting failure with Unbounded page width]` I just pushed and suggest any possible improvements?

Reply:
Makes sense :+1: The fact that you cannot have a `Fail` without having a `Union` prior is quite the hard thing to wrap my head around ^^ I am still thinking about ways this could not be true, but I don't think there are any, because as soon as the algorithm hits fail it returns it, and the only branching code is from `Union` (which made implementing this in a non-lazy language really hard because it's not a tail-call!).